File: BOJ(Java 공부용)/2798.java
import java.io.*;

import static java.lang.Integer.max;

public class Main {
    static int ans = 0;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        String arr[] = br.readLine().split(" ");
        int n= Integer.parseInt(arr[0]),m = Integer.parseInt(arr[1]);
        boolean visited[] = new boolean[n];
        String[] s = br.readLine().split(" ");
        go(0,0,3,s,0,visited,m);
        bw.write(String.valueOf(ans));
        bw.flush();
    }

    private static void go(int idx,int cnt,int depth,String[] s,int total,boolean[] visited,int m) {
        if(cnt==depth) {
            if(total<=m) ans = max(ans,total);
            return;
        }
        for(int i=idx;i<s.length;i++) {
            int num = Integer.parseInt(s[i]);
            if(!visited[i]) {
                visited[i] =true;
                go(idx + 1, cnt + 1, depth, s, total + num,visited,m);
                visited[i] = false;
            }
        }
    }
}

File: BOJ(Kotlin)/2798.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*
import kotlin.math.max
import kotlin.math.min

var ans = 0
fun main(args: Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val st = StringTokenizer(br.readLine())
    val n = st.nextToken().toInt()
    val m = st.nextToken().toInt()
    val StringList : List<String> = br.readLine().split(" ")
    val visited : BooleanArray = BooleanArray(n)
    BlackZack(0,0,3,StringList,0,visited,m)
    bw.write(ans.toString()+"\n")
    bw.flush()
    bw.close()
}

fun BlackZack(idx : Int, cnt : Int, depth : Int, s : List<String>, total : Int, visited : BooleanArray, m : Int) {
    if (cnt == depth) {
        if (total <= m ) ans = max(ans, total)
        return
    }
    for (i in idx until s.size) {
        val num = s.get(i).toInt()
        if (!visited[i]) {
            visited[i] = true
            BlackZack(idx + 1, cnt + 1, depth, s, total + num, visited, m )
            visited[i] = false
        }
    }
}